R/pool.R

```diff
@@ -45,118 +53,178 @@ Pool <- R6Class("Pool",
       }
       reg.finalizer(self,
         function(self) {
-          protectDefaultScheduler({
-            freeEnv <- private$freeObjects
-            if (length(freeEnv) > 0) {
-              self$close()
-            }
-          })
+          if (self$valid) {
+            warning("Closing leaked pool.")
+            self$close()
+          }
         },
         onexit = TRUE)
     },
+
     ## calls activate and returns an object
-    fetch = function() {                       #### WHAT TO DO IF THIS THROWS AN ERROR???
-      protectDefaultScheduler({
+    fetch = function() {
+      if (!self$valid) {
+        stop("This pool is no longer valid. Cannot fetch new objects.")
+      }
+      naiveScheduler$protect({
         if (self$counters$free + self$counters$taken >= self$maxSize) {
           stop("Maximum number of objects in pool has been reached")
         }
 
-        freeEnv <- private$freeObjects
         ## see if there's any free objects
+        freeEnv <- private$freeObjects
         if (length(freeEnv) > 0) {
-          ## get first free object we find
-          id <- ls(freeEnv)[[1]]
+          id <- ls(freeEnv)[[1]]  ## get first free object we find
           object <- freeEnv[[id]]
-
-          taskHandle <- attr(object, "reapTaskHandle", exact = TRUE)
-          if (!is.null(taskHandle)) {
-            attr(object, "reapTaskHandle") <- NULL
-            # Cancel the reap task.
-            taskHandle()
-          }
+          private$cancelReapTask(object)  ## cancel reap task if it exists
 
         } else {
           ## if we get here, there are no free objects
           ## and we must create a new one
           object <- private$createObject()
-          id <- attr(object, "id", exact = TRUE)
-        }
-        ## activate object and return it
-        private$changeObjectStatus(id, object, "free", "taken")
-        onActivate(object)
-        if (!onValidate(object)) {
-          stop("Object activation was not successful")
+          id <- attr(object, "..metadata", exact = TRUE)$..id
         }
 
+        ..metadata <- attr(object, "..metadata", exact = TRUE)
+        naiveScheduler$protect({
+          ## stop recurring validation
+          if (!is.null(..metadata$..stopValHandle)) {
+            ..metadata$..stopValHandle()
+          }
+          private$checkValid(object) ## calls onActivate and onValidate
```

Comment:
You are absolutely right. Sorry about this, I wish I had a unit test for this.

[Aside: I actually thought about making one, but it would be tricky, since from the user's perspective, this would all happen because of a single call to `poolCheckout()`; I could test the functions themselves, but I've been trying to follow [Hadley's advice](http://r-pkgs.had.co.nz/tests.html) of:

> Focus on testing the external interface to your functions - if you test the internal interface, then it’s harder to change the implementation in the future because as well as modifying the code, you’ll also need to update all the tests.

But maybe I should let that go for this situation and actually test the underlying functions...?]

So, I see two ways out of this:
- **the easy way**: just copy the object around and return it (this will either be the same object or the newly created one if validation fails once, but not twice) -- easy, but not efficient.

- **the ugly way**: undo the code refactoring and have all the code back as part of `fetch`. Efficient, but ugly, unreadable and not particularly maintainable.

I lean towards the first one (which is implemented as of right now) because the second one is just so ugly... But none of these seem great to me.

Maybe we could talk about doing something else entirely, like just throwing the first time validation fails and ask the user to try again in the error message itself...?


PS: The `changeObjectStatus` was not supposed to be called from within `checkValid` (I think that must have been a copy-paste error). Removed it now. Good catch, thanks!